containers/src/Kokkos_DynRankView.hpp

```diff
@@ -1110,14 +1099,8 @@ class DynRankView : public ViewTraits<DataType, Properties...> {
             template createLayout<traits, P...>(arg_prop, arg_layout),
         Impl::ViewCtorProp<P...>::has_execution_space);
 
-//------------------------------------------------------------
-#if defined(KOKKOS_ENABLE_CUDA)
-    if (std::is_same<Kokkos::CudaUVMSpace,
-                     typename traits::device_type::memory_space>::value) {
-      typename traits::device_type::memory_space::execution_space().fence(
-          "Kokkos::DynRankView<>::DynRankView: fence after UVM allocation");
-    }
-#endif
+    //------------------------------------------------------------
+    Impl::allocation_fence_if_shared_space<traits>(arg_prop,"Kokkos::DynRankView<>::DynRankView: fence after UVM allocation");
```

Comment:
I would probably have a fence here for all host-accessible allocations. This would cover the host-pinned spaces and shared spaces across all GPU backends and `Kokkos::HostSpace`. Note that the only asynchronous host execution space is `HPX`.

Reply:
Indeed, it would make sense to do so for any host-accessible allocation.

How would you check that? Would something like the following work?

Instead of
```c++
#if defined(KOKKOS_ENABLE_CUDA)
template <typename traits_t, typename alloc_prop_t>
requires std::same_as<Kokkos::CudaUVMSpace,typename traits_t::device_type::memory_space>
constexpr inline void allocation_fence_if_shared_space(alloc_prop_t&& prop, const char* msg)
{ ... }
#endif
```
we would write
```c++
template <typename traits_t, typename alloc_prop_t>
constexpr inline
std::enable_if_t<Kokkos::SpaceAccessibility<Kokkos::Serial, typename traits_t::device_type::memory_space>::accessible>
allocation_fence_if_shared_space(alloc_prop_t&& prop, const char* msg)
{ ... }
```

(I noticed you're not on C++20 yet, so I'll need to remove the `requires` I guess)

Reply:
I would replace
```C++
 if (std::is_same<Kokkos::CudaUVMSpace,
                     typename traits::device_type::memory_space>::value) {
      typename traits::device_type::memory_space::execution_space().fence(
          "Kokkos::DynRankView<>::DynRankView: fence after UVM allocation");
    }
```
with
```C++
 if (Kokkos::SpaceAccessibility<Kokkos::HostSpace, typename traits::device_type::memory_space>::accessible) {
      Impl::get_property<Impl::ExecutionSpaceTag>(prop_copy).fence(
          "Kokkos::DynRankView<>::DynRankView: fence after host-accessible allocation");
    }
```
since `Serial` is not necessarily available. Instead of `HostSpace`, `DefaultHostExecutionSpace` would also work. Also, note that `prop_copy` already contains the execution space instance you want to use.
Personally, I find it clearer to leave the fence inline but maybe you can make a point for creating a separate function. 🙂

Reply:
> I would probably have a fence here for all host-accessible allocations. This would cover the host-pinned spaces and shared spaces across all GPU backends and `Kokkos::HostSpace`. Note that the only asynchronous host execution space is `HPX`.

Also, note that the current approach (fencing if allocation is host-accessible) will also fence if the view is allocated in the host space.

Also, we should change the message attached to the fence (`fence after allocating UVM` is not so relevant anymore). What about `fence after host-accessible memory allocation`?

Reply:
@masterleinad It seems that tests are failingg because, as mentioned above, now a fence is added e.g. when allocations are done in the host memory space. As this was not the case before, tests are failing.

It looks like we have somehow changed the behaviour e.g. for OpemMP.

What do you suggest? Changing the tests expectations or modify the circumstances under which fencing is required after allocation?

Reply:
I would change the tests but we should also see what others think. In the end, we are trying to preserve behavior that is inconsistent with the rest of Kokkos's interfaces. If you give an execution space, the call must be assumed to be asynchronous and the result can only be accessed after fencing.

Reply:
If I understand it correctly, if we pass in a `space`, the view allocation shouldn't fence since, semantically, it's the user's duty to make appropriate fencing when accessing the view? (not sure I got this right)

Reply:
I'll try to bring it up at the next developer meeting.

Reply:
I'm closing this thread as well, because this was discussed below.